File: 2022/cpp/src/3_2.cpp

#include <bits/stdc++.h>
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <queue>
#include <math.h>
#include <map>
#include <set>
#include <unordered_set>
#include <stack>
#include <list>
#include <bitset>
#include <utility>
#include <algorithm>
#include <iomanip>
#include <cstdio>
#include <cstring>
#include <numeric>

using namespace std;

#define all(x) (x).begin(), (x).end()
#define mp make_pair
#define fi first
#define se second
#define pb push_back
#define eb emplace_back
#define forn(i, n) for (int i = 0; i < (int)n; ++i)
#define ford(i, n) for (int i = (int)n - 1; i >= 0; --i)

typedef long long ll;
typedef vector<int> vi;
typedef vector<vi> vvi;
typedef pair<int, int> ii;
typedef vector<ii> vii;
typedef vector<vii> vvii;

const ll mod = 1000000007;
ll gcd(ll a, ll b) { return b ? gcd(b, a % b) : a; }
ll lcm(ll a, ll b) { return a * (b / gcd(a, b)); }

auto inside = [](int row, int col, int r, int c)
{
    return 0 <= row && 0 <= col && row < r && col < c;
};

int main()
{
    // Read file
    ifstream in("../data/in/3.in");

    // HashMap of a-z and A-Z
    map<char, int> priority_map;

    for (int i = 0; i < 26; i++)
    {
        priority_map['a' + i] = i + 1;
        priority_map['A' + i] = i + 27;
    }

    int res = 0;

    // Count map
    map<char, int> count_map_1;
    map<char, int> count_map_2;

    string l1, l2, l3;

    while (!in.eof())
    {
        getline(in, l1), getline(in, l2), getline(in, l3);

        // Count first
        for (int i = 0; i < l1.size(); i++)
        {
            count_map_1[l1[i]]++;
        }

        // Count second
        for (int i = 0; i < l2.size(); i++)
        {
            count_map_2[l2[i]]++;
        }

        // Find in third the one that is also in first and second
        for (int i = 0; i < l3.size(); i++)
        {
            if (count_map_1[l3[i]] > 0 && count_map_2[l3[i]] > 0)
            {
                res += priority_map[l3[i]];
                break;
            }
        }

        count_map_1.clear();
        count_map_2.clear();
    }

    cout << res << endl;
    return 0;
}

File: 2022/python/3_2.py
from itertools import zip_longest

def grouper(iterable, n):
    args = [iter(iterable)] * n
    return zip_longest(*args, fillvalue=None)

f = open("../data/in/3.in", "r")

sum = 0

# HashMap a-z and A-Z
map = dict()

for i in range(26):
    map[chr(i + 97)] = i + 1
    map[chr(i + 65)] = i + 27

count_map_1 = dict()
count_map_2 = dict()

for (l1, l2, l3) in grouper(f.read().splitlines(), 3):
   # Count first
    for ch in l1:
        count_map_1[ch] = count_map_1.get(ch, 0) + 1
    
    # Count second
    for ch in l2:
        count_map_2[ch] = count_map_2.get(ch, 0) + 1
    
    # Find in third the one that is first and second
    for ch in l3:
        if count_map_1.get(ch, 0) > 0 and count_map_2.get(ch, 0) > 0:
            sum += map[ch]
            break

    count_map_1.clear()
    count_map_2.clear()
    
print(sum)